research/delf/delf/python/datasets/sfm120k/sfm120k.py

```diff
@@ -0,0 +1,224 @@
+      qsize: Integer, number of query images.
+      poolsize: Integer, size of the negative image pool, from where the
+        hard-negative images are chosen.
+      loader: Callable, a function to load an image given its path.
+      eccv2020: Bool, whether to use a new validation dataset used with ECCV
+        2020 paper (https://arxiv.org/abs/2007.13172).
+
+    Raises:
+      ValueError: Raised if `mode` is not one of 'train' or 'val'.
+    """
+    if mode not in ['train', 'val']:
+      raise ValueError(
+        "`mode` argument should be either 'train' or 'val', passed as a "
+        "String.")
+
+    # Setting up the paths for the dataset.
+    if eccv2020:
+      name = "retrieval-SfM-120k-val-eccv2020"
+    else:
+      name = "retrieval-SfM-120k"
+    db_root = os.path.join(data_root, 'train/retrieval-SfM-120k')
+    ims_root = os.path.join(db_root, 'ims/')
+
+    # Loading the dataset db file.
+    db_fn = os.path.join(db_root, '{}.pkl'.format(name))
+
+    with tf.io.gfile.GFile(db_fn, 'rb') as f:
+      db = pickle.load(f)[mode]
+
+    # Setting full paths for the dataset images.
+    self.images = [id2filename(img_name, None) for
+                   img_name in db['cids']]
+
+    # Initializing tuples dataset.
+    super().__init__(name, mode, db_root, imsize, nnum, qsize, poolsize,
+                     loader, ims_root)
+
+  def Sfm120kInfo(self):
+    """Metadata for the Sfm120k dataset.
+
+    The dataset contains the image names lists for training and
+    validation, the cluster ID (3D model ID) for each image and indices
+    forming query-positive pairs of images. The images are loaded per epoch
+    and resized on the fly to the desired dimensionality.
+
+    Returns:
+      info: dictionary with the dataset parameters.
+    """
+    info = {'train': {'clusters': 91642, 'pidxs': 181697, 'qidxs': 181697},
+            'val': {'clusters': 6403, 'pidxs': 1691, 'qidxs': 1691}}
+    return info
+
+
+def CreateDataset(mode, data_root, imsize=None, nnum=5, qsize=2000,
+                  poolsize=20000, loader=utils.default_loader,
+                  eccv2020=False):
+  '''Creates Structure-from-Motion (Sfm120k) dataset.
+
+  Args:
+    mode: String, either 'train' or 'val'.
+    data_root: Path to the root directory of the dataset.
+    imsize: Integer, defines the maximum size of longer image side.
+    nnum: Integer, number of negative images per one query.
+    qsize: Integer, number of query images.
+    poolsize: Integer, size of the negative image pool, from where the
+      hard-negative images are chosen.
+    loader: Callable, a function to load an image given its path.
+    eccv2020: Bool, whether to use a new validation dataset used with ECCV
+      2020 paper (https://arxiv.org/abs/2007.13172).
+
+  Returns:
+    sfm120k: Sfm120k dataset instance.
+  '''
+  return _Sfm120k(mode, data_root, imsize, nnum, qsize, poolsize, loader,
+                  eccv2020)
+
+
+def download_train(data_dir):
```

Comment:
Suggestion: let's move this function to a different file in this folder. This sfm120k.py is already quite long, and it would provide a nice separation between functionalities. Maybe it could be called `dataset_download.py`